### ISSUE TITLE
"Appimages bloat the system"

### ISSUE THREAD

**Issue by Samueru-sama** [trytomakeyouprivate/dont-use-appimages#1]
I would like you to fix that statement you have in there that appimages bloat your system, they are actually even smaller than native packages thanks to their compression: 

This the space flatpak takes with just firefox installed: 

![flatpak moment](https://github.com/boredsquirrel/dont-use-appimages/assets/36420837/ce1cfcbe-935b-435f-a9b4-516be98510d1)

This is the space 15 appimages take, that includes heavy applications like libreoffice, kdenlive and 2 web browsers: 

![appimage moment](https://github.com/boredsquirrel/dont-use-appimages/assets/36420837/d744f9f7-ecac-43ee-837a-a2c919622242)

Also I don't know what issue you have with appimages security wise, they are not less safe than portable builds which is what was used instead before appimages, they are also not less than than the Aur and similar methods which are insanely popular (and for a good reason). 

Nobody is forcing you to use appimages, if you don't like them simply keep using flatpak and contribute to it if you haven't. Which if you ask me, I don't like flatpak because it is not terminal app friendly (everything has to be started with `flatpak run org.etc.etc`) and also because flatpak is not XDG Base dir compliant: 

https://github.com/flatpak/flatpak.github.io/issues/191

EDIT: Here is a more proper comparison of the space used by flatpak: 

![flatpak moment2](https://github.com/boredsquirrel/dont-use-appimages/assets/36420837/f95b4403-95a9-4037-bffe-e2216f602296)

**Comment by Samueru-sama** [comment]
>While complex .tar archives (like firefox) may seem messy, they integrate many different things. An installer script takes care of placing a .desktop entry, you can have an updater script, a LICENSE, README and more. Those are all missing with Appimages.

This is totally false, appimages can update themselves (Ferdium for example does it). And there is also nothing stopping the appimage from adding the .desktop entry and icon itself. 

All you would have to do would be add a few cp commands in the `AppRun` script in the appimage that would ask the user if they want to place the desktop entry in `$XDG_DATA_HOME/applications` and  the icon in `$XDG_DATA_HOME/icons/hicolor/...`

Normally you would use tools like AM, Zap, appimagelauncher or even gearlever for the flatpak users that do this for you. And AM even keeps the appimages up to date by comparing the version to that of the repo. 

>Appimages are an insecure and user-unfriendly packaging system with very limited use cases.

They are also very good for quickly troubleshooting a regression in a application, since it involves launching each appimage version rapidly. 

I actually had to use an older version of Ferdium for a while because ferdium was hit by a electron bug that broke it's zoom functionality. It was either that or downgrading electron and manually building ferdium again. 

>Instead users follow strange habits like placing the files on their desktop, which is a highly discouraged "Windows workflow" ([symbolic image](https://www.firstclasstechnologies.co.uk/wp-content/uploads/busy-desktop.png)) and not even supported on many Desktop Environments, most notably [GNOME](https://linuxiac.b-cdn.net/wp-content/uploads/2023/09/gnome45-desktop.jpg).

I use appimage in a tiling wm, what are you even trying to say here. They show up nicely in my rofi menu.

>This means that in the end users need to create their own .desktop entries, that will

>not get updated

That's not the point of portable builds either, and once again you can have appimages that get updated either themselves or thru a 3rd party manager. 

>not get edited when an "update" changes the name of the appimage file (Appimages have no name!)

Same as above. 

> and not deleted if you delete the .appimage file.

This is only true if you add the desktop entry manually, the good thing of this is that desktop entries support `TryExec` which will prevent it from showing in the applications menu if the appimage was removed.

So you only waste a few KiB of disk space if you don't want to bother to remove the desktop entries. 

>(There already was malware [on the Snap store](https://www.bleepingcomputer.com/news/linux/malicious-package-found-on-the-ubuntu-snap-store/).)

The malware in the snap store wasn't because lack of sand boxing, it was because there was no review in the centralized place that distributes them.

Also the recent xz attack sanboxing wouldn't have helped much either when systemd itself was affected...

>Please just use Flatpak, save us all the trouble and support a modern Linux desktop.

What trouble are you saving by not using appimages? appimages a basically an extra step from portable builds which are commonly released already by developers. And save the hassle of having to uncompress the tar file and run the binary manually for the user.

In other words, you take the portable build and add the `AppRun` mainly just calls the binary inside, a desktop entry and a icon. That is too much for the devs?

You're essentially telling devs to stop releasing their binaries and force people to deal with a format that can't be converted back to appimage and others as easily... 

>([Firejail is not recommended](https://madaidans-insecurities.github.io/linux.html#firejail)).

That very link also mentions that flatpak isn't good for sanboxing either. And even calls out the flatpak devs for saying that sandboxing with Xpra is impossible kek.

**Comment by Samueru-sama** [comment]
Also somebody on a certain website said that I wasn't fair with flatpak because flatpak has deduplication of libraries and it is only the first application that makes it big. 

So I just installed kdenlive and libreoffice alongside firefox, and this is the size of flatpak now: 

BEHOLD, DEDUPLICATION: 

![image](https://github.com/boredsquirrel/dont-use-appimages/assets/36420837/235cb11d-b12c-4ce4-8897-8b17dfb312b8)

How can this shit be so bad, libreoffice alone increased the size by 1.4 GiB, **while the uncompressed libreoffice appimage is 860 MiB.** (the compressed appimage is 323.8 MiB btw). 

EDIT: I decided to actually put the flatpaks in their own partition and check the partition size instead: 

```
Alright I just moved flatpak to its own partition and checked the size of the partition instead:

with firefox, kdenlive and libreoffice:

Disk (/var/lib/flatpak) 2.69 GiB / 19.12 GiB (14%) - ext4

That’s much better now. But still twice the size that 15 appimages took.

This is with now having firefox librewolf brave kdenlive and libreoffice:

Disk (/var/lib/flatpak) 3.40 GiB / 19.12 GiB (18%) - ext4

Still though, the appimages take less space. A by a large margin.
```

It is not the world ending levels that my file manager was reporting, but still with 5 application flatpak is taking near 3X as much as appimages would.

--------------------------------------------

EDIT: Here is a more proper comparison of the space used by flatpak: 

![flatpak moment2](https://github.com/boredsquirrel/dont-use-appimages/assets/36420837/f95b4403-95a9-4037-bffe-e2216f602296)

Reminder that 15 appimages is 1.2 GIB again...

**Comment by IverCoder** [comment]
The space most GUI file managers report is excessive because Flatpak makes heavy use of deduplication and linking, resulting on many files ending up being counted twice. This is on top of most libraries being decoupled through runtimes.

You can read more details about it [here](https://blogs.gnome.org/wjjt/2021/11/24/on-flatpak-disk-usage-and-deduplication/).

**Comment by IverCoder** [comment]
> What trouble are you saving by not using appimages? appimages a basically an extra step from portable builds which are commonly released already by developers. And save the hassle of having to uncompress the tar file and run the binary manually for the user.
> 
> In other words, you take the portable build and add the AppRun mainly just calls the binary inside, a desktop entry and a icon. That is too much for the devs?
> 
> You're essentially telling devs to stop releasing their binaries and force people to deal with a format that can't be converted back to appimage and others as easily...

You are missing the point on ease of use. A normal person unlike you will not want to have to create seperate .desktop files just to be able to get an app show nicely on the app list. Flatpak handles everything for us, unlike in AppImage where we need to do it ourselves unnecessarily.

**Comment by Samueru-sama** [comment]
> The space most GUI file managers report is excessive because Flatpak makes heavy use of deduplication and linking, resulting on many files ending up being counted twice. This is on top of most libraries being decoupled through runtimes.
> 
> You can read more details about it [here](https://blogs.gnome.org/wjjt/2021/11/24/on-flatpak-disk-usage-and-deduplication/).

Yeah that is why I updated the post with screenshots using the partition size and also `flatpak-dedup-checker`. It is where I say "EDIT: Here is a more proper comparison of the space used by flatpak:"

>You are missing the point on ease of use

![image](https://github.com/boredsquirrel/dont-use-appimages/assets/36420837/6309d017-0c49-40e6-8003-7ee3d5ad7ecc)


EDIT: I also took a screenshot using flatpak dedup checker with just firefox installed in flatpak, forgot to put it here: 

![image](https://github.com/boredsquirrel/dont-use-appimages/assets/36420837/2bab0cdf-4b95-457f-a2a2-60a0af39f730)

(It is still using more than 15 appimages for a single application).

**Comment by IverCoder** [comment]
The space advantages would be apparent when you use more and more Flatpaks. I have **over a hundred** Flatpak apps installed, including heavy ones such as LibreOffice, and my home folder usage (excluding Steam games) is less than 15 GiB. Over **a hundred apps**, plus all my documents yet less than 15 GiB of disk usage??? That's some serious disk space efficiency.N

Now, let's compare if I had over a hundred of AppImages. Each and every one of them bundles practically every library they need short of `glibc` with no means of deduplicating them. My laptop's disk space would be dead long before I can download all my apps.

I can't even get started about how hypocrytical it is to hear AppImage users complaining about Flatpak's disk usage.

**Comment by Samueru-sama** [comment]
> The space advantages would be apparent when you use more and more Flatpaks. I have **over a hundred** Flatpak apps installed, including heavy ones such as LibreOffice, and my home folder usage (excluding Steam games) is less than 15 GiB. Over **a hundred apps**, plus all my documents yet less than 15 GiB of disk usage??? That's some serious disk space efficiency.N
> 
> Now, let's compare if I had over a hundred of AppImages. Each and every one of them bundles practically every library they need short of `glibc` with no means of deduplicating them. My laptop's disk space would be dead long before I can download all my apps.
> 

My entire distro + 20 appimages + the`home user files is 5.2 GiB... Which the 20 appimages is 1.5 GiB. 

![image](https://github.com/boredsquirrel/dont-use-appimages/assets/36420837/ea35d7dd-e71e-4a79-ab02-cbe47b925abf)

Doing a crude calculation, 100 appimages would be about 7.5 GiB.

> I can't even get started about how hypocrytical it is to hear AppImage users complaining about Flatpak's disk usage.

I've showed you clear evidence that demonstrates that appimages use less space and by a large margin.

**Comment by boredsquirrel** [comment]
I dont currently have the time to update the post. But yes, I agree the situation is not black/white.

I have to agree though, that even though 3rd party tools exist, the upstream developers mostly never endorse them, so people will not use them automatically.

A system needs to work for tech-incompetent people, and do that securely. There are many reasons why this is not possible.

**Comment by Samueru-sama** [comment]
Btw when you have time also clear that statement that appimages need fuse2, there is go-appimage/appimagetool that builds appimages that work with fuse3. I use it for all my appimages and also cemu ships their appimage with it as well. 

And iirc go-appimage also statically links musl so it should be possible to make appimages that work in alpine, but I don't know if one exists yet. 

But yeah, it is true that the big appimages like libreoffice and kdenlive still build with the older appimagetool and as result need fuse2, the good news is that those can be updated to work with fuse3 with tools like AM and all you have to do is run `am nolibfuse kdenlive` for example since you don't need fuse2 to extract the appimage to convert it. None of this requires sudo either (besides needing to install fuse3 if isn't in the distro).

**Comment by boredsquirrel** [comment]
so...

- [ ] appimages dont use deduplication but only pull in needed libraries, making them often smaller in practice
- [ ] there is a tool for building appimages with fuse3, it is only used by cemu but you could rebuild them. it is also written in Go, which is a plus

**Comment by Samueru-sama** [comment]
> so...
> 
> * [ ]  appimages dont use deduplication but only pull in needed libraries, making them often smaller in practice
> * [ ]  there is a tool for building appimages with fuse3, it is only used by cemu but you could rebuild them. it is also written in Go, which is a plus

Is the Go comment a joke or what lol. 

amd_gpu-top and other appimages build with cargo-appimage also work with fuse3 already btw. I only know about amdgpu_top and cemu because I use them.

**Comment by ibnuwadyain** [comment]
i use appimages in my chromebook linux , i can put icons on my desktop very simply by just putting the appimages themselves or just symlinks in my "/usr/bin" directory, the apps automatically added in the launcher dock with its icon, so no need to add desktop icons either manually or using 3rd party apps anymore.
appimages are perfect for the chromebooks limited storage space capacity, the appimages can be put and run from attached pen drives and symlinked as in the trick above.

**Comment by Samueru-sama** [comment]
> i use appimages in my chromebook linux , i can put icons on my desktop very simply by just putting the appimages themselves or just symlinks in my "/usr/bin" directory, the apps automatically added in the launcher dock with its icon, so no need to add desktop icons either manually or using 3rd party apps anymore. appimages are perfect for the chromebooks limited storage space capacity, the appimages can be put and run from attached pen drives and symlinked as in the trick above.

That's good but I have a small suggestion to what you are doing: 

Instead of putting symlinks in `/usr/bin` create `$HOME/.local/bin` (which is part of the XDG Base dir spec) and put/link your appimages there. That `$HOME/.local/bin` can also just be a symlink to your flash drive as well. And make sure that location is part of your `$PATH`, which your distro should be doing that already but many (like arch and forks) don't. 

To do so simply add `export PATH="$HOME/.local/bin:$PATH"` to your bash_profile or zprofile, etc

None of this requires elevated privileges btw. 

Also the fact that your appimages show in your launcher dock means that there is some daemon running that is integrating them I think.
